src/fal_planner/execute.py

```diff
@@ -0,0 +1,46 @@
+import time
+from concurrent.futures import FIRST_COMPLETED, ThreadPoolExecutor, wait
+from typing import List
+
+from fal_planner.plan import TaskQueue, load_graph
+
+N_THREADS = 5
+
+
+def serial_executor(queue: TaskQueue) -> None:
+    while task := queue.get_next_task():
+        task.execute()
+        queue.finish(task)
+
+
+def thread_based_executor(queue: TaskQueue) -> None:
+    with ThreadPoolExecutor(max_workers=N_THREADS) as executor:
+        futures = {
+            executor.submit(task.execute) for task in queue.iter_available_tasks()
+        }
+        while futures:
+            # Get the first completed futures, mark them done.
+            done, futures = wait(futures, return_when=FIRST_COMPLETED)
```

Comment:
is this always returning a single future or list of futures?

Reply:
It returns two set of futures, one is done and the other is pending/running. The done set might contain one item (since we use `FIRST_COMPLETED`), but there are no guarantees AFAIK (e.g., if while we are working on this two futures finishes in the second iteration we might receive both of them [didn't test tho]).